docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+
+### Recommended number of nodes and zones
+
+It is recommended that for high-availability setup an odd number of nodes (node tolerance) or zones (zone tolerance) must be used. This also follows the [recommendations](https://etcd.io/docs/v3.4/faq/#why-an-odd-number-of-cluster-members) on the etcd cluster size. The recommendations for number of zones will be largely influenced by quorum-based etcd cluster setup recommendations as other shoot control plane components are either stateless or non-quorum-based stateful components.
+
+**Let's take the following example to explain this recommendation further:**
+* Seed clusters' worker nodes are spread across **two zones**
+* Gardener would distribute a **three member** etcd cluster - AZ-1: 2 replicas, AZ-2: 1 replica
+* If AZ-1 goes down then, quorum is lost and the only remaining etcd member enters into a read-only state.
+* If AZ-2 goes down then:
+  * If the leader is in AZ-2, then it will force a re-election and the quorum will be restored with 2 etcd members in AZ-1.
+  * If the leader is not in AZ-2, then etcd cluster will still be operational without any downtime as the quorum is not lost.
+
+**Result:**
+* There seems to be no clear benefit to spreading an etcd cluster across 2 zones as there is an additional cost of cross-zonal traffic that will be incurred due to communication amongst the etcd members and also due to API server communication with an etcd member across zones.
+* There is no significant gain in availability as compared to an etcd cluster provisioned within a single zone. Therefore it is a recommendation that for regions having 2 availability zones, etcd cluster should only be spread across nodes in a single AZ.
+
+### Recommended number of replicas
+
+The minimum number of replicas required to achieve HA depends on the [topology](#topologies) and the requirement of each component that run in an `active-active` mode.
+
+**Active-Active**
+* If application needs a quorum to operate (e.g. `etcd`), at least **three replicas** are required ([ref](https://etcd.io/docs/v3.3/faq/#why-an-odd-number-of-cluster-members)).
+* Non-quorum based components are also supposed to run with a minimum count of **three replicas** to survive node/zone outages and support load balancing.
+
+**Active-Passive**
+* Components running in  a `active-passive` mode are expected to have at least **two** replicas, so that there is always one replica on stand-by.
+
+## Gardener Shoot API
+
+### Proposed changes
+
+The following changes to the shoot API are suggested to enable the HA feature for a shoot cluster:
+
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  controlPlane:
+    highAvailability:
+      enabled: true # by default HA is disabled.
+      failureDomain:
+        type: <single-zone | multi-zone>
+        whenUnsatisfied: <DoNotSchedule | ScheduleAnyway>
```

Comment:
You have a point here and the GEP should also initiate an adequate discussion on it if we should rather use the explicit and simple way as you explained or by allowing some dynamic configuration with `whenUnsatisfied`.

Backgroup and coming back to you analogy description:

There is a feature called [Pod Topology Spread Constraint](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/) which exactly provides a `whenUnsatisfied` feature.

The idea was to offer the same functionality with `single-/multi-zone` as well.

Reply:
To add to what @timuthy has already mentioned, it is similar to `preferredDuringSchedulingIgnoredDuringExecution` vs `requiredDuringSchedulingIgnoredDuringExecution` that you also find as an option for affinity rules. Hosting HA shoot control plane will require an availability of an HA seed. If no such seed is available that can satisfy the multi-zone spread of the HA control plane then should we fail the provisioning of the shoot or should we still provide you HA but with a single zonal spread.

If all seed clusters tomorrow are going to be multi-zonal then this point would be moot. But if that is not going to be the case then it could still provide flexibility to the end user to define what should be the behavior.

Reply:
For now as @timuthy has suggested we can drop this from the spec.

Reply:
Okay, thanks for explaining. I didn't know that the `whenUnsatisfied` fields comes from [Pod Topology Spread Constraint](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/) and its idea is similar to it. The field was just not explained in the document and I could only derive from other comments what is the idea behind it.

---

> Hosting HA shoot control plane will require an availability of an HA seed. If no such seed is available that can satisfy the multi-zone spread of the HA control plane then should we fail the provisioning of the shoot or should we still provide you HA but with a single zonal spread.

For the initial version I would start small and pick the "fail the provisioning and ask the end use to update its failureDomain type or let the gardener operators add a new multi-zone Seed".

I would to consider adding support for `whenUnsatisfied` only when there is need/demand for it. WDYT?

Reply:
yes that is why we have removed this option from the shoot spec in this iteration.